pkg/resource/deploy/plan.go

```diff
@@ -0,0 +1,612 @@
+
+	environmentVariables := make(map[string]string)
+	for _, e := range os.Environ() {
+		// golang has so many questionable design decisions, lets just hope no one puts '=' in there env var names.
+		pair := strings.SplitN(e, "=", 2)
+		environmentVariables[pair[0]] = pair[1]
+	}
+
+	return Plan{
+		ResourcePlans:        make(map[resource.URN]*ResourcePlan),
+		EnvironmentVariables: environmentVariables,
+		Manifest:             manifest,
+		Config:               config,
+	}
+}
+
+// Clone makes a deep copy of the given plan and returns a pointer to the clone.
+func (plan *Plan) Clone() *Plan {
+	return copystructure.Must(copystructure.Copy(plan)).(*Plan)
+}
+
+// PlanDiff holds the results of diffing two object property maps.
+type PlanDiff struct {
+	Adds    resource.PropertyMap   // the resource's properties we expect to add.
+	Deletes []resource.PropertyKey // the resource's properties we expect to delete.
+	Updates resource.PropertyMap   // the resource's properties we expect to update.
+}
+
+// Returns true if the Deletes array contains the given key
+func (planDiff *PlanDiff) ContainsDelete(key resource.PropertyKey) bool {
+	found := false
+	for i := range planDiff.Deletes {
+		if planDiff.Deletes[i] == key {
+			found = true
+			break
+		}
+	}
+	return found
+}
+
+func (planDiff *PlanDiff) MakeError(
+	key resource.PropertyKey,
+	actualOperation string,
+	actualValue *resource.PropertyValue) string {
+
+	// diff wants to do 'actualOperation' (one of '+', '~', '-', '=') but plan differs. This function looks up what
+	// key wanted to do to print a more useful error message
+
+	// See if the plan was an add, remove, update, or nothing to give a better error message
+	var expectedOperation string
+	var expectedValue *resource.PropertyValue
+	if expected, has := planDiff.Adds[key]; has {
+		expectedValue = &expected
+		expectedOperation = "+"
+	} else if expected, has = planDiff.Updates[key]; has {
+		expectedValue = &expected
+		expectedOperation = "~"
+	} else if planDiff.ContainsDelete(key) {
+		expectedOperation = "-"
+	} else {
+		expectedOperation = "="
+	}
+	diff := ""
+	if actualValue != nil && expectedValue != nil {
+		diff = "[" + expectedValue.String() + "!=" + actualValue.String() + "]"
+	} else if actualValue != nil {
+		diff = "[" + actualValue.String() + "]"
+	} else if expectedValue != nil {
+		diff = "[" + expectedValue.String() + "]"
+	}
+	return expectedOperation + actualOperation + string(key) + diff
+}
+
+// Goal is a desired state for a resource object.  Normally it represents a subset of the resource's state expressed by
+// a program, however if Output is true, it represents a more complete, post-deployment view of the state.
+type GoalPlan struct {
+	// the type of resource.
+	Type tokens.Type
+	// the name for the resource's URN.
+	Name tokens.QName
+	// true if this resource is custom, managed by a plugin.
+	Custom bool
+	// the resource's checked input properties that we saw during preview.
```

Comment:
Ah looks like here the doc comments are different between CheckedInputs and InputDiff.